### PR TITLE
Improve string representations of entities

### PR DESCRIPTION
**Pull request by smurching**
Currently, MLflow entities use default Python string serialization, resulting in hard-to-interpret output like:

```
>>> from  mlflow.tracking import get_service
>>> service = get_service()
>>> experiments = service.list_experiments()
[<mlflow.entities.experiment.Experiment at 0x7f93a8902a90>]
```

This PR improves entities' string representations to include a listing of their fields (and recursive serialization of lists of entities, e.g. lists of metrics/params within a run). The above now results in:

```
>>> from  mlflow.tracking import get_service
>>> service = get_service()
>>> experiments = service.list_experiments()
[<mlflow.entities.experiment.Experiment: 'artifact_location'='/Users/sid/code/mlflow/mlruns/0', 'experiment_id'=0, 'name'='Default'>]
```

Runs now appear as:
```
>>> r = service.get_run(run_id)
>>> r
<mlflow.entities.run.Run: 'data'=<mlflow.entities.run_data.RunData: 'metrics'=[<mlflow.entities.metric.Metric: 'key'='random_int', 'timestamp'=1534987517, 'value'=75.0>, <mlflow.entities.metric.Metric: 'key'='foo', 'timestamp'=1534987517, 'value'=7.0>], 'params'=[<mlflow.entities.param.Param: 'key'='param1', 'value'='5'>], 'tags'=[]>, 'info'=<mlflow.entities.run_info.RunInfo: 'artifact_uri'='/Users/sid/code/mlflow/mlruns/0/145d84e1b29c4fe9a3b191f27e63a79f/artifacts', 'end_time'=1534987517756, 'entry_point_name'='', 'experiment_id'=0, 'name'='Run 6', 'run_uuid'='145d84e1b29c4fe9a3b191f27e63a79f', 'source_name'='example/advanced/test_remote_server.py', 'source_type'=4, 'source_version'='e375c52679ab74b3f6e864d6bcab0118125e31e4', 'start_time'=1534987517633, 'status'=3, 'user_id'='sid'>>
```

Feedback is welcome - run serialization is still somewhat unwieldy due to the number of fields in a run.

### CODE REVIEW

**Review by mparkhe — COMMENTED** [review]
Very cool implementation. Minor comment about property ordering in output.

File: mlflow/entities/_mlflow_object.py [hunk 1]
```diff
@@ -20,3 +21,37 @@ def from_proto(cls, proto):
     @classmethod
     def from_dictionary(cls, the_dict):
         return cls(**the_dict)
+
+    def __repr__(self):
+        return serialize(self)
+
+
+def serialize(obj):
```
Comment (mateiz) [inline mlflow/entities/_mlflow_object.py]: Just as a note, I think "serialize" can mean converting to formats other than string representation, so I would call this to_string or pretty_print or something.

File: mlflow/entities/_mlflow_object.py [hunk 2]
```diff
@@ -20,3 +21,37 @@ def from_proto(cls, proto):
...
+        # Handle nested lists inside MLflow entities (e.g. lists of metrics/params)
+        if isinstance(obj, list):
+            res = [serialize(elem) for elem in obj[:self._MAX_LIST_LEN]]
+            if len(obj) > self._MAX_LIST_LEN:
+                res.append("...")
+            return "[%s]" % ", ".join(res)
+        return self.printer.pformat(obj)
+
+    def serialize_entity(self, entity):
+        return ", ".join(["%s=%s" % (self.serialize(key), self.serialize(value))
```
[the remark this answers is not in the document]
Comment (mateiz) [reply mlflow/entities/_mlflow_object.py]: +1, that would be nice. Also don't put quotes around the key -- basically change the `self.serialize(key)` to just `key`.

File: mlflow/entities/_mlflow_object.py [hunk 3]
```diff
@@ -20,3 +21,37 @@ def from_proto(cls, proto):
...
+class _MLflowObjectPrinter(object):
+    _MAX_LIST_LEN = 2
+
+    def __init__(self):
+        super(_MLflowObjectPrinter, self).__init__()
+        self.printer = pprint.PrettyPrinter()
+
+    def serialize(self, obj):
+        if isinstance(obj, _MLflowObject):
+            return "<%s: %s>" % (get_classname(obj), self.serialize_entity(obj))
```
Comment (mateiz) [inline mlflow/entities/_mlflow_object.py]: I'd include just the last part of the class name: Experiment instead of mlflow.entities.experiment.Experiment.

**Comment by smurching** [comment]
Thanks @mateiz @mparkhe, merging to master!

**Pull request merged**